src/copy.c

```diff
@@ -840,6 +929,19 @@ copyfrom(CopyChunkState *ccstate, List *range_table, Hypertable *ht, MemoryConte
 				(errmsg("Using normal unbuffered copy operation (CIM_SINGLE) "
 						"because triggers are defined on the destination table.")));
 	}
+	else if (Max(ts_guc_max_open_chunks_per_insert, ts_guc_max_cached_chunks_per_hypertable) <
+			 MIN_CHUNKS_FOR_MULTI_COPY)
```

Comment:
I'm not sure we should allow such low values of these caches at all, besides for testing. Maybe we should ereport a warning here.

Reply:
Another argument against this is that it makes testing stuff such as reusing insert buffers much more difficult.

Reply:
The copy code has two code paths: 

* `CIM_SINGLE` for chunks that use features (e.g., triggers) that require a sequential order for insertion. 
* `CIM_MULTI_CONDITIONAL` is used when multi-inserts are possible. 

So, we have had two code paths so far, which are used depending on the configuration of the table.

This new condition ensures that `CIM_SINGLE` is used when only a few chunks can be opened (and cached) in parallel. The overhead for maintaining the multi-insert buffer is higher in this case than using direct bulk inserts (see attached graphs). 

The condition is deterministic and can be tested by setting the parameters in the test properly. In the test, you can easily check whether multi-buffer optimization is used or not by setting the log level to DEBUG1. We also already have tests that verify the used code path in that way.

Reply:
Let me expand on the test coverage stuff. Say, I want to test re-using insert buffer for a different chunk. Now I can set `ts_guc_max_open_chunks_per_insert = 1` and insert two lines. But if we disable CIM_MULTI for small number of chunks, I'll have to make a bigger data set for this. Why not allow it with a warning? Such low values are not a valid configuration anyway, but useful for tests.